wicket/src/inventory.rs

```diff
@@ -20,94 +21,70 @@ impl Inventory {
         self.power.get(id)
     }
 
-    pub fn update_power_state(
-        &mut self,
-        id: ComponentId,
-        state: PowerState,
-    ) -> anyhow::Result<()> {
-        Self::validate_component_id(id)?;
-        self.power.insert(id, state);
-        Ok(())
-    }
-
     pub fn get_inventory(&self, id: &ComponentId) -> Option<&Component> {
         self.inventory.get(id)
     }
 
     pub fn update_inventory(
         &mut self,
-        id: ComponentId,
-        component: Component,
+        inventory: RackV1Inventory,
     ) -> anyhow::Result<()> {
-        Self::validate_component_id(id)?;
-        self.inventory.insert(id, component);
-        Ok(())
-    }
-
-    fn validate_component_id(id: ComponentId) -> anyhow::Result<()> {
-        match id {
-            ComponentId::Sled(i) if i > 31 => {
-                Err(anyhow!("Invalid sled slot: {}", i))
-            }
-            ComponentId::Switch(i) if i > 1 => {
-                Err(anyhow!("Invalid switch slot: {}", i))
-            }
-            ComponentId::Psc(i) if i > 1 => {
-                Err(anyhow!("Invalid power shelf slot: {}", i))
-            }
-            _ => Ok(()),
+        let mut new_inventory = Inventory::default();
+
+        for sp in inventory.sps {
+            let i = sp.id.slot;
+            let type_ = sp.id.type_;
+            let sp = Sp { ignition: sp.ignition, state: sp.state };
+
+            // Validate and get a ComponentId
+            let (id, component) = match type_ {
+                SpType::Sled => {
+                    if i > 31 {
+                        return Err(anyhow!("Invalid sled slot: {}", i));
+                    }
+                    (ComponentId::Sled(i as u8), Component::Sled(sp))
+                }
+                SpType::Switch => {
+                    if i > 1 {
+                        return Err(anyhow!("Invalid switch slot: {}", i));
+                    }
+                    (ComponentId::Switch(i as u8), Component::Switch(sp))
+                }
+                SpType::Power => {
+                    if i > 1 {
+                        return Err(anyhow!("Invalid power shelf slot: {}", i));
+                    }
+                    (ComponentId::Psc(i as u8), Component::Psc(sp))
+                }
+            };
+            new_inventory.inventory.insert(id, component);
```

Comment:
It's a little weird that `Inventory` has a field also named `inventory`, I think? This question might not make sense in a larger context, but why do we want to record the component info and power state in separate fields?

Reply:
Ah, so this is just a historical artifact. I wanted to plug in the real inventory without changing the UI at this time. I think it maybe worth a rework at some point, possibly as I pull in real power.